Framework/DataObjects/src/ScanningWorkspaceBuilder.cpp

```diff
@@ -0,0 +1,323 @@
+
+  m_indexingType = indexingType;
+}
+
+/**
+ * Verify everything has been set that is required and return the workspace.
+ *
+ * @return Workspace2D with the scanning information set
+ */
+MatrixWorkspace_sptr ScanningWorkspaceBuilder::buildWorkspace() {
+  validateInputs();
+
+  auto outputWorkspace = create<Workspace2D>(
+      m_instrument, m_nDetectors * m_nTimeIndexes, m_histogram);
+
+  auto &outputDetectorInfo = outputWorkspace->mutableDetectorInfo();
+  outputDetectorInfo.setScanInterval(0, m_timeRanges[0]);
+
+  buildOutputDetectorInfo(outputDetectorInfo);
+
+  if (!m_positions.empty())
+    buildPositions(outputDetectorInfo);
+
+  if (!m_rotations.empty())
+    buildRotations(outputDetectorInfo);
+
+  if (!m_instrumentAngles.empty())
+    buildInstrumentAngles(outputDetectorInfo);
+
+  switch (m_indexingType) {
+  case IndexingType::DEFAULT:
+    outputWorkspace->setIndexInfo(
+        Indexing::IndexInfo(m_nDetectors * m_nTimeIndexes));
+    break;
+  case IndexingType::TIME_ORIENTED:
+    createTimeOrientedIndexInfo(*outputWorkspace);
+    break;
+  case IndexingType::DETECTOR_ORIENTED:
+    createDetectorOrientedIndexInfo(*outputWorkspace);
+    break;
+  }
+
+  return boost::shared_ptr<MatrixWorkspace>(std::move(outputWorkspace));
+}
+
+void ScanningWorkspaceBuilder::buildOutputDetectorInfo(
+    DetectorInfo &outputDetectorInfo) {
+  for (size_t i = 1; i < m_nTimeIndexes; ++i) {
+    const auto mergeWorkspace =
+        create<Workspace2D>(m_instrument, m_nDetectors, m_histogram.binEdges());
+    auto &mergeDetectorInfo = mergeWorkspace->mutableDetectorInfo();
+    for (size_t j = 0; j < m_nDetectors; ++j) {
+      mergeDetectorInfo.setScanInterval(j, m_timeRanges[i]);
+    }
+    outputDetectorInfo.merge(mergeDetectorInfo);
+  }
+}
+
+void ScanningWorkspaceBuilder::buildRotations(
+    DetectorInfo &outputDetectorInfo) const {
+  for (size_t i = 0; i < m_nDetectors; ++i) {
+    for (size_t j = 0; j < m_nTimeIndexes; ++j) {
+      outputDetectorInfo.setRotation({i, j}, m_rotations[i][j]);
+    }
+  }
+}
+
+void ScanningWorkspaceBuilder::buildPositions(
+    DetectorInfo &outputDetectorInfo) const {
+  for (size_t i = 0; i < m_nDetectors; ++i) {
+    for (size_t j = 0; j < m_nTimeIndexes; ++j) {
+      outputDetectorInfo.setPosition({i, j}, m_positions[i][j]);
+    }
+  }
+}
+
+void ScanningWorkspaceBuilder::buildInstrumentAngles(
+    DetectorInfo &outputDetectorInfo) const {
+  for (size_t i = 0; i < outputDetectorInfo.size(); ++i) {
+    for (size_t j = 0; j < outputDetectorInfo.scanCount(i); ++j) {
+      auto position = outputDetectorInfo.position({i, j});
+      const auto rotation =
+          Kernel::Quat(m_instrumentAngles[j], Kernel::V3D(0, 1, 0));
+      rotation.rotate(position);
+      outputDetectorInfo.setPosition({i, j}, position);
+      outputDetectorInfo.setRotation({i, j}, rotation);
+    }
+  }
+}
+
+void ScanningWorkspaceBuilder::createTimeOrientedIndexInfo(
```

Comment:
Also, since all these such functions are private, consider making these free functions and keeping the function declarations out of the header file. These can just live in an anonymous namespace.